src/compiler.ts

```diff
@@ -6991,16 +7127,74 @@ export class Compiler extends DiagnosticEmitter {
   // <reference-counting>
 
   /** Makes a retain call, retaining the expression's value. */
-  makeRetain(expr: ExpressionRef): ExpressionRef {
+  makeRetain(
+    expr: ExpressionRef,
+    type: Type | null = null
```

Comment:
RC is now dependent on type- here's the new process:
- If the expression represents a function reference, which are now managed, we check to see if it's a Closure using the MSB, then retain/release the pointer (which needs to be shifted by 4 bits to get the 16-bit aligned pointer). For retaining, we also need to set this value back to the local we are retaining, so we need to shift the bits/reset the MSB before returning
- Otherwise, if the type isn't passed in (calls which could not reference a closure) or if the type isn't a function (a type represented by a signature only- closures which are represented by signatures and class references don't count. They are normal pointers and don't need this special treatment), so we just retain/release as normal

Reply:
Updated process:
- Get either the closure pointer if the MSB set, or 0 otherwise.
- Release/Retain that result

Reply:
Wondering if this can be `makeRetain(expr: ExpressionRef, isFunction: bool = false)`, so we end up with less code changes?

Reply:
Well we would still need to alter most of the makeRetain statements to include this new isFunction parameter, so it won't decrease the diff that way. Unless I'm missing something and it's impossible for these retain statements to be targeting a closure (and there are a lot of them, so I should probably look through again), in which case we can just omit the type on those ones, which would reduce the diff